src/redis_memory.rs

```diff
@@ -0,0 +1,242 @@
+    Address: Hash + Eq,
+{
+    connection: Arc<Mutex<redis::Connection>>,
+    _marker: PhantomData<Address>,
+}
+
+impl<Address: Hash + Eq, const WORD_LENGTH: usize> Debug for RedisMemory<Address, WORD_LENGTH> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RedisMemory")
+            .field("connection", &"<redis::Connection>") // We don't want to debug the actual connection
+            .field("_marker", &self._marker)
+            .finish()
+    }
+}
+
+impl<Address: Hash + Eq, const WORD_LENGTH: usize> Default for RedisMemory<Address, WORD_LENGTH> {
+    fn default() -> Self {
+        Self {
+            connection: match redis::Client::open("redis://127.0.0.1:9999/") {
+                Ok(client) => match client.get_connection() {
+                    Ok(con) => Arc::new(Mutex::new(con)),
+                    Err(e) => {
+                        panic!("Failed to connect to Redis: {}", e);
+                    }
+                },
+                Err(e) => panic!("Error creating redis client: {:?}", e),
+            },
+            _marker: PhantomData,
+        }
+    }
+}
+
+/**
+ * Flushes the Redis database.
+ * WARNING: This is irreversible, do not run in production.
+ */
+#[cfg(test)]
+impl<Address: Hash + Eq + Debug, const WORD_LENGTH: usize> RedisMemory<Address, WORD_LENGTH> {
+    pub fn flush_db(&self) -> Result<(), redis::RedisError> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        redis::cmd("FLUSHDB").exec(safe_connection)?;
+        Ok(())
+    }
+}
+
+/**
+ * The RedisMemory implementation of the MemoryADT trait.
+ * All operations are - and should be - atomic.
+ */
+impl<Address: Send + Sync + Hash + Eq + Debug + Clone + ToRedisArgs, const WORD_LENGTH: usize>
+    MemoryADT for RedisMemory<Address, WORD_LENGTH>
+{
+    type Address = Address;
+    type Error = RedisMemoryError;
+    type Word = [u8; WORD_LENGTH];
+
+    /**
+     * Atomically reads the values at the given addresses.
+     */
+    async fn batch_read(
+        &self,
+        addresses: Vec<Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        let refs: Vec<&Address> = addresses.iter().collect(); // Redis MGET requires references to the values
+        let res: Vec<Option<Self::Word>> =
+            safe_connection.mget(&refs).map_err(|e| RedisMemoryError {
+                details: e.to_string(),
+            })?;
+        Ok(res)
+    }
+
+    /**
+     * Atomically writes the bindings if the guard is still valid.
+     * Returns the current value on the guard's address.
+     * If the result is equal to the guard's old value, the bindings get
+     * written.
+     *
+     * Args that are passed to the LUA script are, in order :
+     * 1. Guard address.
+     * 2. Guard value.
+     * 3. Vector length.
+     * 4+. Vector elements (address, word).
+     */
```

Comment:
This function is already documented in the trait. Plus your doc is wrong as you do not return "the current value of the guard's address".

Reply:
I kept this part because the new lua script is not so evident to read, is it ok ?
**/// Args that are passed to the LUA script are, in order:
/// 1. Guard address.
/// 2. Guard value. 
/// 3. Vector length.
/// 4+. Vector elements (address, word).**

and deleted the rest

Reply:
no because this will replace the trait documentation. Plus these are implementation details and have no use in the API doc. You can either document your script (which you have extracted as an independent constant) or write them as comment (for the developer's attention)